pkg/traces/traces.go

```diff
@@ -0,0 +1,99 @@
+package traces
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"reflect"
+	"runtime"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	semconv "go.opentelemetry.io/otel/semconv/v1.17.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	ApplicationName           = "launcher"
+	defaultSpanName           = "launcher/unknown"
+	defaultAttributeNamespace = "unknown"
+)
+
+// StartSpan returns a new context and span, including information about the calling function
+// as appropriate. Standardizes the tracer name. The caller is always responsible for
+// ending the span. `keyVals` should be a list of pairs, where the first in the pair is a
+// string representing the attribute key and the second in the pair is the attribute value.
+func StartSpan(ctx context.Context, keyVals ...interface{}) (context.Context, trace.Span) {
+	spanName := defaultSpanName
+
+	opts := make([]trace.SpanStartOption, 0)
+
+	// Extract information about the caller to set some standard attributes (code.filepath,
+	// code.lineno, code.function) and to set more specific span and attribute names.
+	// runtime.Caller(0) would return information about `StartSpan` -- calling
+	// runtime.Caller(1) will return information about the function calling `StartSpan`.
+	pc, callerFile, callerLine, ok := runtime.Caller(1)
+	if ok {
+		opts = append(opts, trace.WithAttributes(
+			semconv.CodeFilepath(callerFile),
+			semconv.CodeLineNumber(callerLine),
+		))
+
+		// Extract the calling function name and use it to set code.function and the span name.
+		if f := runtime.FuncForPC(pc); f != nil {
+			spanName = filepath.Base(f.Name())
+			opts = append(opts, trace.WithAttributes(semconv.CodeFunction(f.Name())))
+		}
+	}
+
+	opts = append(opts, trace.WithAttributes(buildAttributes(callerFile, keyVals...)...))
+
+	return otel.Tracer(ApplicationName).Start(ctx, spanName, opts...)
+}
+
+// buildAttributes takes the given keyVals, expected to be pairs representing the key
+// and value of each attribute, and parses them appropriately, ensuring that the keys
+// have consistent and specific names. Pairs that cannot be parsed will be discarded.
+func buildAttributes(callerFile string, keyVals ...interface{}) []attribute.KeyValue {
+	callerDir := defaultAttributeNamespace
+	if callerFile != "" {
+		callerDir = filepath.Base(filepath.Dir(callerFile))
```

Comment:
I don't know what this value is -- is the path separator going to vary by platform?

Reply:
It's the closest I got to getting the package name -- e.g. when the caller is ee/localserver/request-controlservice.go, this will be `localserver`. I can document this better.

Reply:
Added a comment to make this clearer

Reply:
On windows, are the paths `/` or `\`? (I'm never sure, since it's oddly inconsistent. windows paths are, of course `\`, but sometimes the internal go paths are `/`)

Reply:
Oh sorry, I didn't realize I'd missed that part of the question -- let me confirm.

Reply:
Forward slash also -- so no, path separator will not vary by platform.

![Screenshot 2023-06-09 122315](https://github.com/kolide/launcher/assets/8119675/58d1f9c7-4bf4-4da6-8fec-682a027fe2d5)

Reply:
Then it'll need `path` not `filepath` and a comment.